### PR TITLE
connector_importer: misc fix/imp

### PR DESCRIPTION
**Pull request by damdam-s**
[REF] create a base context for create/write method
[IMP] add a "_after_import" feature
[IMP] add a feature to set values in the shared storage at the end of the import
[FIX] management of the "_xml_id::module.name" notation
[FIX] import_record parser of options now works with DotDict
[REM] remove old "settings" notation
[IMP] better management of datetime format in utils convert tool

### CODE REVIEW

**Comment by damdam-s** [comment]
@simahawk some fixes/changes (non breaking)

File: connector_importer/components/importer.py [hunk 1]
```diff
@@ -272,13 +272,20 @@ def _load_mapper_options(self):
         return {"override_existing": self.recordset.override_existing}
 
     # TODO: make these contexts customizable via recordset settings
+    def _odoo_default_context(self):
+        """Default context to be used in both create and write methods"""
+        return {
+            "importer_type": self.recordset.import_type_id.id,
```
Comment (simahawk) [inline connector_importer/components/importer.py]:
```suggestion
            "importer_type_id": self.recordset.import_type_id.id,
```

File: connector_importer/components/importer.py [hunk 2]
```diff
@@ -342,6 +350,7 @@ def run(self, record, is_last_importer=True, **kw):
                             continue
                         odoo_record = self.record_handler.odoo_create(values, line)
                         self.tracker.log_created(values, line, odoo_record)
+                    self._after_import(odoo_record, values, line)
```
Comment (simahawk) [inline connector_importer/components/importer.py]: this can be done on the recordhandler no?

File: connector_importer/models/record.py [hunk 3]
```diff
@@ -71,14 +72,14 @@ def debug_mode(self):
...
         """This job will import a record.
 
-        # TODO rewrite
-        :param component_name: name of the importer component to use
-        :param model_name: name of the model to import
-        :param is_last_importer: flag for last importer of the recordset
+        :param importer_config: configuration of the importer
         """
+        # set the importer_config as DotDict
+        importer_config = DotDict(importer_config_dict)
```
Comment (simahawk) [inline connector_importer/models/record.py:81]: when the dict arrives here should be already a DotDict

**Review by simahawk — CHANGES_REQUESTED** [review]
@damdam-s tnx for your contrib. Can you please split each change in its own commit?

File: connector_importer/models/record.py (hunk 3 above)
Comment (damdam-s) [reply connector_importer/models/record.py:81]: that's not the case and I was not able to run the importer because type was `dict`. maybe the serialization/unserialization has something to do here

File: connector_importer/components/importer.py (hunk 2 above)
Comment (damdam-s) [reply connector_importer/components/importer.py]: yeah probably. I was expecting to use it later but that's not the case anymore. I'll remove it

File: connector_importer/models/record.py (hunk 3 above)
Comment (simahawk) [reply connector_importer/models/record.py:81]: yeah, let me rephrase: please make sure we get always a DotDict before we get here :wink:

**Comment by ivantodorovich** [comment]
Both pre-commit and travis are red, could you check?

**Comment by simahawk** [comment]
and split commits squashed together on the 1st one :pray:

File: connector_importer/components/importer.py [hunk 4]
```diff
@@ -363,9 +370,13 @@ def run(self, record, is_last_importer=True, **kw):
             ]
         ).format(**counters)
         self.tracker._log(msg)
+        self._set_shared_storage_values(self.tracker.chunk_report)
         self._trigger_finish_events(record, is_last_importer=is_last_importer)
         return counters
 
+    def _set_shared_storage_values(self, chunk_report):
```
Comment (simahawk) [inline connector_importer/components/importer.py:377]: this is not setting anything. I'd prefer to keep it generic and rename it to `_finished_chunk_hook()`.
Just call `self._finished_chunk_hook()`, no need to pass the chunk_report as you have it in `self.tracker`.
Or maybe just pass the last line and odoo record if any.

Comment (simahawk) [reply connector_importer/components/importer.py:377]: Actually: any reason why you cannot use the event hook instead?

File: connector_importer/components/mapper.py [hunk 5]
```diff
@@ -81,7 +83,11 @@ def default_values(self, record=None):
         values = {}
         for k, v in self.defaults:
             if isinstance(v, str) and v.startswith("_xmlid::"):
-                xmlid, field_value = v.split("::")[1:]
+                xml_id_value_list = v.split("::")[1:]
+                if not xml_id_value_list:
+                    raise exceptions.UserError(_("Malformated xml id ref"))
+                xml_id_value = xml_id_value_list[0]
+                xmlid, field_value = xml_id_value.split(":")
```
Comment (simahawk) [inline connector_importer/components/mapper.py:90]:
```suggestion
                __, value = v.partition("::")
                if not value:
                    raise exceptions.UserError(_("Malformed xml id ref"))
                xmlid, field_value = value.split(":")
```

File: connector_importer/models/record.py (hunk 3 above)
Comment (simahawk) [reply connector_importer/models/record.py:81]: this should come only from https://github.com/OCA/connector-interfaces/blob/14.0/connector_importer/models/import_type.py#L120 and as such, we should always have DotDict here.
If not, then we probably need to call it at the end of `_make_importer_info` instead.

**Review by simahawk — CHANGES_REQUESTED** [review]
Looks better, thanks! Some more remarks :innocent:

**Comment by simahawk** [comment]
@damdam-s I moved to https://github.com/OCA/connector-interfaces/pull/90 ready/accepted commits and rewrote them a bit.
I excluded:

1. [[REM] remove old settings notation](https://github.com/OCA/connector-interfaces/pull/86/commits/d140835361e53393dd48e447689ff4d5fc960da6) because if you do it now you have to provide an upgrade script -> feel free to add it :pray: 
2. [[IMP] add a feature to set values in the shared storage at the end of…](https://github.com/OCA/connector-interfaces/pull/86/commits/36b8450bacc911e7797ff9a86d83bb74f9961f73) because I'm not sure we need it.

So, for now I close this PR. Feel free to reopen it once is ready again, thanks! :pray: 

Meanwhile, you can maybe do a quick review on #90 :wink:
